vcd/resource_vcd_nsxt_route_advertisement.go

```diff
@@ -0,0 +1,168 @@
+	"log"
+	"strings"
+)
+
+func resourceVcdNsxtRouteAdvertisement() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceVcdNsxtRouteAdvertisementCreateUpdate,
+		ReadContext:   resourceVcdNsxtRouteAdvertisementRead,
+		UpdateContext: resourceVcdNsxtRouteAdvertisementCreateUpdate,
+		DeleteContext: resourceVcdNsxtRouteAdvertisementDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceVcdNsxtRouteAdvertisementImport,
+		},
+		Schema: map[string]*schema.Schema{
+			"org": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Description: "The name of organization to use, optional if defined at provider " +
+					"level. Useful when connected as sysadmin working across different organizations",
+			},
+			"vdc": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "The name of VDC to use, optional if defined at provider level",
+				Deprecated:  "Edge Gateway will be looked up based on 'edge_gateway_id' field",
+			},
+			"edge_gateway_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "NSX-T Edge Gateway ID in which route advertisement is located",
+			},
+			"enabled": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+				Description: "Define if route advertisement is active",
+			},
+			"subnets": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Set of subnets that will be advertised to Tier-0 gateway. Leaving it empty means none",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func resourceVcdNsxtRouteAdvertisementCreateUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+	vcdClient.lockParentEdgeGtw(d)
+	defer vcdClient.unLockParentEdgeGtw(d)
+
+	var subnets []string
+	enableRouteAdvertisement := d.Get("enabled").(bool)
+	subnetsFromSchema, subnetsArgumentSet := d.GetOk("subnets")
+
+	if subnetsArgumentSet {
+		subnets = convertSchemaSetToSliceOfStrings(subnetsFromSchema.(*schema.Set))
+	}
+
+	if !enableRouteAdvertisement && len(subnets) > 0 {
+		return diag.Errorf("if enable is set to false, no subnets must be passed")
+	}
+
+	_, edgeGateway, err := getParentEdgeGatewayOwnerIdAndNsxtEdgeGateway(vcdClient, d, "route advertisement")
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	_, err = edgeGateway.UpdateNsxtRouteAdvertisement(enableRouteAdvertisement, subnets, true)
+	if err != nil {
+		return diag.Errorf("error when creating/updating route advertisement - %s", err)
+	}
+
+	d.SetId(edgeGateway.EdgeGateway.ID)
+
+	return resourceVcdNsxtRouteAdvertisementRead(ctx, d, meta)
+}
+
+func resourceVcdNsxtRouteAdvertisementRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	orgName := d.Get("org").(string)
```

Comment:
I changed for the helper function you said.
The thing is, doing it that way is a little inefficient since we are retrieving the whole Org just to get its name, but I agree its much safer

Reply:
You can add better or additional helper function, or deal with provider lvl ORG also.